spyder/plugins/ipythonconsole/widgets/main_widget.py

```diff
@@ -2220,20 +2210,35 @@ def norm(text):
             client = self.get_current_client()
 
         if client is not None:
-            # Internal kernels, use runcell
-            if client.get_kernel() is not None and not run_cell_copy:
+            is_spyder_kernel = client.shellwidget.is_spyder_kernel
+            # Only use copy for runcell
+
+            if method == 'runcell' and (run_cell_copy or not is_spyder_kernel):
+                # Use copy of cell
+                line = code.strip()
+            elif is_spyder_kernel:
+                # Use custom function
                 line = (str(
                         "{}({}, '{}')").format(
-                                str(function),
+                                str(method),
                                 repr(cell_name),
                                 norm(filename).replace("'", r"\'")))
 
-                if function == "debugcell":
+                if method == "debugcell":
                     # To keep focus in editor after running debugfile
                     client.shellwidget._pdb_focus_to_editor = focus_to_editor
             # External kernels and run_cell_copy, just execute the code
             else:
-                line = code.strip()
+                # Can not use custom function on non-spyder kernels
+                QMessageBox.warning(
+                    self,
+                    _('Warning'),
+                    _("The client is not a spyder-kernel "
+                      "to run <b>{}</b>.<br><br>Please Use a spyder-kernel."
+                      ).format(method),
+                    QMessageBox.Ok
+                )
+                return
```

Comment:
Why this change? If a kernel is not a Spyder one, we can still run code on it by simply pasting it, which is what I think we are doing here, aren't we?

Reply:
Yes but not debug, so if it is not a spyder and not tu file, we can not run